File: 19-countingSundays/sundays.c
#include <stdio.h>
#include <stdbool.h>

typedef struct date {
	int year;
	int month;
	int day;
	int week_day;
} Date;

bool is_leap_year(int year) {
	if ( year % 4 == 0 ) {
		if ( year % 100 == 0 ) {
			if ( year % 400 == 0 ) {
				return true;
			}
			return false;
		}

		return true;
	}

	return false;
}

bool is_thirty_day_month(int month) {
	switch(month) {
		case 4:
		case 6:
		case 9:
		case 11:
			return true;
		default:
			return false;
	}
}

void process_month(Date * current) {
	if ( current->month == 2 ) {
		if ( is_leap_year(current->year) && current->day == 30 ) {
			current->day = 1;
			current->month++;
		} else if ( current->day == 29 ) {
			current->day = 1;
			current->month++;
		}
	} else if ( is_thirty_day_month(current->month) ) {
		if ( current->day == 31 ) {
			current->day = 1;
			current->month++;
		}
	} else if ( current->day == 32 ) {
		current->day = 1;
		current->month++;
	}

	return;
}

int main(void) {
	Date date = { 1900, 1, 1, 2 }; 

	bool looping  = true;
	bool counting = false;

	int sundays = 0;

	while (looping) {
		// exit conditions
		if ( date.year >= 1901 ) {
			counting = true;
		}
		if ( date.day == 31 && date.month == 12 && date.year == 2000 ) {
			counting = false;
			looping  = false;
			continue;
		}

		date.day++;
		date.week_day++;

		process_month(&date);

		if ( date.week_day == 8 ) {
			date.week_day = 1;
		}

		if ( date.month == 13 ) {
			date.year++;
			if ( is_leap_year(date.year) ) {
				printf("%i is a leap year!\n", date.year);
			}
			date.month = 1;
		}

		if ( date.day == 1 && date.week_day == 1 && counting ) {
			sundays++;
			printf("Sunday -> %i-%02i-%02i [%i]\n", date.year, date.month, date.day, sundays);
		}
	}

	printf("\nFinished! Total Sundays = %i\n\n", sundays);

	return 0;
}

File: 19-countingSundays/sundays.js
// sundays.js 
// Eric Ryan Harrison (2023) | https://github.com/blister 

/*
 * 1 Jan 1900 = Monday
 * 28 Days = Feb (year divisible by 4 except centuries divisible by 400)
 * 30 Days = September, April, June, November
 * 31 Days = All others
 *
 * how many sundays fell on the first of the month between 1 Jan 1901 -> 31 Dec 2000)
 */

let week_day = 2; // sunday = 1, mon = 2, sat = 7
let day = 1;
let month = 1;
let year = 1900;

const thirty = {4:true, 6:true, 9:true, 11:true};

let sundays = 0;

let looping = true;
let counting = false;
while ( looping ) {
	// only count sundays between 1 Jan 1901 and 31 Dec 2000
	if ( year >= 1901 ) {
		counting = true;
	}
	if ( day === 31 && month === 12 && year === 2000 ) {
		looping = false;
		continue;
	}

	day++;
	week_day++;

	// feb
	if ( month === 2 && day > 28 ) {
		if ( year % 4 !== 0 ) {
			
		}

	}

	// 30 day months
	if ( month in thirty ) {
		
	}
}

// sundays.js, created: 2023-227-15 09:35

